### ISSUE TITLE
File Not Found Because The Path Changes?

### ISSUE THREAD

**Issue by perlmunger** [FolioReader/FolioReaderKit#190]
This is a weird one. It doesn't happen with all .epubs, but the one I am working with has this issue where the path I get from the bundle (loading it as a local resource) it different than the one the FREpubParser is attempting to parse. In this code:

```swift
override func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
    let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "PubCellIdentifier", for: indexPath) as! PubCollectionViewCell
    if let bookPath = Bundle.main.url(forResource: "iks", withExtension: "epub") {
        if let image = FolioReader.getCoverImage(bookPath.path) {
            cell.imageView.image = image
        }
    }

    return cell
}
```

I break on the line where `getCoverImage` is before that code gets called and I `po` the `bookPath` variable which tells me the file path is at:

> /Users/mlong/Library/Developer/CoreSimulator/Devices/E9FDFDBD-F86F-4B5F-BD57-948114704C8D/data/Containers/Bundle/Application/AC0A7A58-D56D-4516-8000-7E623B6E0C29/CMReader.app/iks.epub

But then I let the code run and it crashes here in the FREpubParser:

<img width="690" alt="frepubparser_swift" src="https://cloud.githubusercontent.com/assets/17556/21037404/b77a5750-bd8a-11e6-994f-b7437d101260.png">

The error is:

> Error excluding iks.epub from backup Error Domain=NSCocoaErrorDomain Code=4 "The file “iks.epub” doesn’t exist." UserInfo={NSURL=file:///Users/mlong/Library/Developer/CoreSimulator/Devices/E9FDFDBD-F86F-4B5F-BD57-948114704C8D/data/Containers/Data/Application/6A607D84-9853-4409-90C5-94D816446B87/Documents/iks.epub/, NSFilePath=/Users/mlong/Library/Developer/CoreSimulator/Devices/E9FDFDBD-F86F-4B5F-BD57-948114704C8D/data/Containers/Data/Application/6A607D84-9853-4409-90C5-94D816446B87/Documents/iks.epub, NSUnderlyingError=0x61800004b9d0 {Error Domain=NSPOSIXErrorDomain Code=2 "No such file or directory"}}
> Cannot read container.xml

If you look at the two file locations, there is a difference for some reason:

```
/Users/mlong/Library/Developer/CoreSimulator/Devices/E9FDFDBD-F86F-4B5F-BD57-948114704C8D/data/Containers/Bundle/Application/AC0A7A58-D56D-4516-8000-7E623B6E0C29/CMReader.app/iks.epub
/Users/mlong/Library/Developer/CoreSimulator/Devices/E9FDFDBD-F86F-4B5F-BD57-948114704C8D/data/Containers/Data/Application/6A607D84-9853-4409-90C5-94D816446B87/Documents/iks.epub
```
(Might be easier to notice in this screenshot:)

<img width="681" alt="diff" src="https://cloud.githubusercontent.com/assets/17556/21037490/6cb150c4-bd8b-11e6-83da-1bd28d9e5cc3.png">

Like I said, it doesn't happen on all files, but it is happening on this one that I need to test with. Any idea why the path is different at the crash?

### Update

I was able to duplicate this with the example project. I embedded this same epub that I'm testing with and got a similar issue with the mismatched epub paths:

```
/Users/mlong/Library/Developer/CoreSimulator/Devices/E9FDFDBD-F86F-4B5F-BD57-948114704C8D/data/Containers/Bundle/Application/4C530AE1-AB83-471D-8D6F-BA941765AA2A/Example.app/iks.epub
/Users/mlong/Library/Developer/CoreSimulator/Devices/E9FDFDBD-F86F-4B5F-BD57-948114704C8D/data/Containers/Data/Application/2420C949-283F-42B8-95FD-0B0301B3D6BF/Documents/iks.epub
```

However, in the example app, the crash is happening when trying to set the skip backup attribute:

<img width="756" alt="skipbackup" src="https://cloud.githubusercontent.com/assets/17556/21037937/06d7c4dc-bd8f-11e6-87b1-d915bc0a6a80.png">

**Comment by hebertialmeida** [comment]
One think I noticed is that seems that it is entering on catch while reading container.xml.
```
Cannot read container.xml
```
Please make sure that this ePub has a valid container at this path `META-INF/container.xml`. I am aware that I need to make this `readEpub` function failable. It can not continue if the container do not exists.

Is it possible for you to send this ePub to my email, so I can test that?

**Comment by hebertialmeida** [comment]
This is probably related to #164

**Comment by perlmunger** [comment]
Thanks for taking a look, Herberti. I appreciate it. 

I'm not familiar enough yet with epubs internal workings. Is the container something you have to add while authoring? Also, I noticed from within the filesystem (terminal) that the one I'm using is actually a directory (bundle) while the ones you have in the sample app are single files--not directories. What is the difference between the two? Is one a legacy format and the other newer? How can I ensure that the epubs have this container you're referring to?

Thanks again!

p.s. I've sent the epub I'm working with to your email address.

**Comment by perlmunger** [comment]
I'm not completely sure, but I'm wondering if the difference between my files is ePub 2 vs ePub 3. I've gone back into InDesign and re-exported the ePub ensuring it's ePub 3 which seems to be the single file rather than the bundle and that seems to be working.

**Comment by hebertialmeida** [comment]
I have checked you ePub and couple things:
- It is a fixed layout ePub, we do not support that yet;
- As you said it is not an single file, probably unzipped with the same file name;
- Yeah I will fix the bug of crashing related to issue #164

**Comment by hebertialmeida** [comment]
Closed due to inactivity.
